packaging/snapd.mk

```diff
@@ -0,0 +1,190 @@
+libexecdir:=$(shell rpm -E %{_libexecdir})
+mandir:=$(shell rpm -E %{_mandir})
+datadir:=$(shell rpm -E %{_datadir})
+localstatedir:=$(shell rpm -E %{_localstatedir})
+unitdir=$(shell rpm -E %{_unitdir})
+else
+$(error expected RPM to be used)
+prefix=/usr
+bindir=$(prefix)/bin
+sbindir=$(prefix)/sbin
+libexecdir=$(prefix)/lib
+mandir=$(prefix)/
+datadir=$(prefix)/share
+localstatedir=/var
+unitdir=$(prefix)/lib/systemd/system
+endif
+
+# Ensure that none of the directory variables are empty.
+vars=bindir sbindir libexecdir mandir datadir localstatedir unitdir
+$(foreach var,$(vars),$(if $(value $(var)),,$(error $(var) is empty or unset)))
+
+# This is usually set by %make_install. It is defined here to avoid warnings or
+# errors from referencing undefined variables.
+DESTDIR?=
+
+# Decide which of the two snap mount directories to use. This is not
+# referencing localstatedir because the code copes with only those two values
+# explicitly.
+ifeq ($(with_alt_snap_mount_dir),1)
+snap_mount_dir=/var/lib/snapd/snap
+else
+snap_mount_dir=/snap
+endif
+
+# The list of go binaries we are expected to build.
+go_binaries=snap snapctl snap-seccomp snap-update-ns snap-exec snapd
+
+# NOTE: This *depends* on building out of tree. Some of the built binaries
+# conflict with directory names in the tree.
+.PHONY: all
+all: $(go_binaries) 
+
+snap snapctl snap-seccomp:
+	go build -buildmode=pie $(import_path)/cmd/$@
+
+# Those two need to be built as static binaries. They run on the inside of a
+# nearly-arbitrary mount namespace that does not contain anything we can depend
+# on (no standard library, for example).
+snap-update-ns snap-exec:
+	go build -buildmode=default -ldflags '-extldflags "-static"' $(import_path)/cmd/$@
+
+# Snapd can be built with test keys. This is only used by the internal test
+# suite to add test assertions. Do not enable this in distribution packages.
+snapd:
+ifeq ($(with_test_keys),1)
+	go build -buildmode=pie -tags withtestkeys $(import_path)/cmd/$@
+else
+	go build -buildmode=pie $(import_path)/cmd/$@
+endif
+
+# Know how to create certain directories.
+$(addprefix $(DESTDIR),$(libexecdir)/snapd $(bindir) $(mandir)/man8 $(localstatedir)/lib/snapd $(localstatedir)/cache/snapd $(snap_mount_dir)):
+	install -d -m 755 $@
+
+.PHONY: install
+
+# Install snap into /usr/bin/.
+install:: snap | $(DESTDIR)$(bindir)
+	install -m 755 $^ $|
+
+# Install snapctl snapd, snap-{exec,update-ns,seccomp} into /usr/lib/snapd/
+install:: snapctl snapd snap-exec snap-update-ns snap-seccomp | $(DESTDIR)$(libexecdir)/snapd
+	install -m 755 $^ $|
+
+# Ensure /usr/bin/snapctl is a symlink to /usr/lib/snapd/snapctl
+install:: | $(DESTDIR)$(bindir)
+	ln -s $(libexecdir)/snapd/snapctl $|/snapctl
+
+# Generate and install man page for snap command
+install:: snap | $(DESTDIR)$(mandir)/man8
+	./snap help --man > $|/snap.8
+
+# Install the directory structure in /var/lib/snapd
+install::
+	install -d $(DESTDIR)$(localstatedir)/lib/snapd/apparmor/profiles
```

Comment:
Might as well set some default permissions for the directories.

Reply:
Hmm? Do you mean `-m 755`

Reply:
Yes